Clean up > 2-year-old features

psarna opened the pull request:
Following the work started in 253a7640e, a new batch of old features is assumed to be always available. They are all still announced via gossip, but the code assumes that the feature is always true, because we only support upgrades from a previous release, and the release window is considerably smaller than 2 years.

Features picked this time via `git blame`, along with the date of their introduction:

* fe4afb1aa3 (Asias He                  2018-09-05 14:52:10 +0800  109) static const sstring ROW_LEVEL_REPAIR = "ROW_LEVEL_REPAIR";
* ff5e541335 (Calle Wilund              2019-02-05 13:06:07 +0000  110) static const sstring TRUNCATION_TABLE = "TRUNCATION_TABLE";
* fefef7b9eb (Tomasz Grabiec            2019-03-05 19:08:07 +0100  111) static const sstring CORRECT_STATIC_COMPACT_IN_MC = "CORRECT_STATIC_COMPACT_IN_MC";

Tests: unit(dev)

haaawk commented:
Two of those features (MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES) are controlled by a flag. Have we removed the flag to make it impossible to turn off those features?

elcallio commented:
lgtm

psarna commented:
> Two of those features (MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES) are controlled by a flag. Have we removed the flag to make it impossible to turn off those features?

No, good catch, the config flags effectively became ignored. I'll change their status to unused - removing them entirely would probably be too harsh, since people could experience failed boots during upgrades. The unused config options deserve a separate garbage collection process as well :)

psarna commented:
Hm, these flags are also used in a couple of tests in manual/ and perf/, so I'll also verify if everything works fine after these flags are no longer taken into account.

avikivity commented on repair/repair.cc:
```diff
@@ -934,211 +933,6 @@ repair_neighbors repair_info::get_repair_neighbors(const dht::token_range& range
-}
-
 // Repair a single local range, multiple column families.
 // Comparable to RepairSession in Origin
```
@asias say good-bye to this code

avikivity commented on repair/repair.cc:
```diff
@@ -934,211 +933,6 @@ repair_neighbors repair_info::get_repair_neighbors(const dht::token_range& range
-                // there are any differences, sync the content of this range.
-                std::vector<future<partition_checksum>> checksums;
-                checksums.reserve(1 + neighbors.size());
-                checksums.push_back(checksum_range(ri.db, ri.keyspace, cf, range, checksum_type));
```
Surely, more code can be dropped - checksum_range for example, and the verbs in messaging_service.

@asias please follow up with more code destruction.

asias replied:
I will do the remaining after this patch set.

avikivity commented on gms/feature_service.cc:
```diff
@@ -45,11 +45,11 @@ constexpr std::string_view features::LA_SSTABLE = "LA_SSTABLE_FORMAT";
 constexpr std::string_view features::STREAM_WITH_RPC_STREAM = "STREAM_WITH_RPC_STREAM";
 constexpr std::string_view features::MC_SSTABLE = "MC_SSTABLE_FORMAT";
 constexpr std::string_view features::ROW_LEVEL_REPAIR = "ROW_LEVEL_REPAIR";
+constexpr std::string_view features::TRUNCATION_TABLE = "TRUNCATION_TABLE";
```
It was used, see 30a700c5b0eb425983ef424bfe3fc87d04043448.

avikivity replied:
Better add this to the commit message.

avikivity commented on sstables/sstables_manager.cc:
```diff
@@ -62,8 +62,7 @@ sstable_writer_config sstables_manager::configure_writer(sstring origin) const {
-    cfg.correctly_serialize_static_compact_in_mc =
-            bool(_features.cluster_supports_correct_static_compact_in_mc());
+    cfg.correctly_serialize_static_compact_in_mc = true;
     cfg.origin = std::move(origin);
```
Please follow up with removal of the member variable, which can no longer vary.

psarna replied:
Yeah, I left it as is for now, because some tests validated what happens if the variable is false. But it's actually better to purge the variable right now (as well as the one above, left from the previous series), I'll do it in v2.

avikivity replied:
Looks like you forgot this.

psarna replied:
oh, right. In v3 then

avikivity commented on gms/feature_service.cc:
```diff
@@ -47,11 +47,11 @@ constexpr std::string_view features::MC_SSTABLE = "MC_SSTABLE_FORMAT";
 // Up-to-date features
 constexpr std::string_view features::UDF = "UDF";
 constexpr std::string_view features::MD_SSTABLE = "MD_SSTABLE_FORMAT";
-constexpr std::string_view features::UNBOUNDED_RANGE_TOMBSTONES = "UNBOUNDED_RANGE_TOMBSTONES";
```
Note: this feature depends on mc sstables (which the series mandates, so no problem).

psarna commented:
Oh, queued already - if it passes next, then I'll prepare all the requests - purging the additional values which are always true + making the config value that enables mc format obsolete - as a follow-up.

avikivity commented:
> Two of those features (MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES) are controlled by a flag. Have we removed the flag to make it impossible to turn off those features?

The flag exists, but is no longer consulted.

haaawk commented:
> > Two of those features (MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES) are controlled by a flag. Have we removed the flag to make it impossible to turn off those features?
> 
> The flag exists, but is no longer consulted.

There's a deeper implication. This PR creates an impression that it just removes features that has been always on for many releases but this is not true for MC_SSTABLE_FORMAT. It is entierly possible that some users have decided to stay on LA sstables. Removing the MC_SSTABLE_FORMAT feature removes this ability. It is at least something that should be mentioned in the release notes (@tzach ). The question also is whether this is a conscious decision or an accidental result.

psarna commented:
> > > Two of those features (MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES) are controlled by a flag. Have we removed the flag to make it impossible to turn off those features?
> > 
> > 
> > The flag exists, but is no longer consulted.
> 
> There's a deeper implication @avikivity. This PR creates an impression that it just removes features that has been always on for many releases but this is not true for MC_SSTABLE_FORMAT. It is entierly possible that some users have decided to stay on LA sstables. Removing the MC_SSTABLE_FORMAT feature removes this ability. It is at least something that should be mentioned in the release notes (@tzach ). The question also is whether this is a conscious decision or an accidental result.

@avikivity that's a valid concern though, theoretically there may be users that want to continue using la.

psarna commented:
But wait, I haven't actually removed the option to turn off mc in config, so it should work fine. I think I just got confused.

denesb commented:
> > > Two of those features (MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES) are controlled by a flag. Have we removed the flag to make it impossible to turn off those features?
> > 
> > 
> > The flag exists, but is no longer consulted.
> 
> There's a deeper implication @avikivity. This PR creates an impression that it just removes features that has been always on for many releases but this is not true for MC_SSTABLE_FORMAT. It is entierly possible that some users have decided to stay on LA sstables. Removing the MC_SSTABLE_FORMAT feature removes this ability. It is at least something that should be mentioned in the release notes (@tzach ). The question also is whether this is a conscious decision or an accidental result.

Why is removing this feature removing this ability? The feature merely means that the cluster is *capable* of using MC and says nothing about whether it is actually using it.

psarna commented:
> > > > Two of those features (MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES) are controlled by a flag. Have we removed the flag to make it impossible to turn off those features?
> > > 
> > > 
> > > The flag exists, but is no longer consulted.
> > 
> > 
> > There's a deeper implication @avikivity. This PR creates an impression that it just removes features that has been always on for many releases but this is not true for MC_SSTABLE_FORMAT. It is entierly possible that some users have decided to stay on LA sstables. Removing the MC_SSTABLE_FORMAT feature removes this ability. It is at least something that should be mentioned in the release notes (@tzach ). The question also is whether this is a conscious decision or an accidental result.
> 
> Why is removing this feature removing this ability? The feature merely means that the cluster is _capable_ of using MC and says nothing about whether it is actually using it.

True. For a while I panicked that I removed the capability to use LA, but it's fortunately not the case (: I agree with the above.

haaawk commented:
> > > Two of those features (MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES) are controlled by a flag. Have we removed the flag to make it impossible to turn off those features?
> > 
> > 
> > The flag exists, but is no longer consulted.
> 
> There's a deeper implication. This PR creates an impression that it just removes features that has been always on for many releases but this is not true for MC_SSTABLE_FORMAT. It is entierly possible that some users have decided to stay on LA sstables. Removing the MC_SSTABLE_FORMAT feature removes this ability. It is at least something that should be mentioned in the release notes (@tzach ). The question also is whether this is a conscious decision or an accidental result.

Actually, removing those MC_SSTABLE_FORMAT is just unsafe. In the rolling upgrade when cluster has MC_SSTABLE_FORMAT disabled, we may end up in a state where some nodes use LA and others MC. When we add UNBOUNDED_RANGE_TOMBSTONES that can't be handled by LA we will end up with troubles. Maybe the unbounded tombstone will be lost or maybe it will be wrongly persisted. I don't know.

The correct way to remove MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES is to remove the flag that controls them in one release and only in the next remove those features. Or am I missing something @avikivity ?

haaawk commented:
> > > > Two of those features (MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES) are controlled by a flag. Have we removed the flag to make it impossible to turn off those features?
> > > 
> > > 
> > > The flag exists, but is no longer consulted.
> > 
> > 
> > There's a deeper implication @avikivity. This PR creates an impression that it just removes features that has been always on for many releases but this is not true for MC_SSTABLE_FORMAT. It is entierly possible that some users have decided to stay on LA sstables. Removing the MC_SSTABLE_FORMAT feature removes this ability. It is at least something that should be mentioned in the release notes (@tzach ). The question also is whether this is a conscious decision or an accidental result.
> 
> Why is removing this feature removing this ability? The feature merely means that the cluster is _capable_ of using MC and says nothing about whether it is actually using it.

Currently users may decide to write LA sstables in their clusters. I don't know why they would do that but it's certainly an option. After this PR all sstables will be written as MC. Ability of reading LA sstables is not affected.

psarna commented:
> > > > Two of those features (MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES) are controlled by a flag. Have we removed the flag to make it impossible to turn off those features?
> > > 
> > > 
> > > The flag exists, but is no longer consulted.
> > 
> > 
> > There's a deeper implication. This PR creates an impression that it just removes features that has been always on for many releases but this is not true for MC_SSTABLE_FORMAT. It is entierly possible that some users have decided to stay on LA sstables. Removing the MC_SSTABLE_FORMAT feature removes this ability. It is at least something that should be mentioned in the release notes (@tzach ). The question also is whether this is a conscious decision or an accidental result.
> 
> Actually, removing those MC_SSTABLE_FORMAT is just unsafe. In the rolling upgrade when cluster has MC_SSTABLE_FORMAT disabled, we may end up in a state where some nodes use LA and others MC. When we add UNBOUNDED_RANGE_TOMBSTONES that can't be handled by LA we will end up with troubles. Maybe the unbounded tombstone will be lost or maybe it will be wrongly persisted. I don't know.
> 
> The correct way to remove MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES is to remove the flag that controls them in one release and only in the next remove those features. Or am I missing something @avikivity ?

We should never be able to upgrade from a cluster that doesn't have MC_SSTABLE_FORMAT enabled **and** this series is applied, because we only support direct upgrades from the previous release, not really old ones. That's why we only deprecate features that are over 2 years old.

psarna commented:
> > > > > Two of those features (MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES) are controlled by a flag. Have we removed the flag to make it impossible to turn off those features?
> > > > 
> > > > 
> > > > The flag exists, but is no longer consulted.
> > > 
> > > 
> > > There's a deeper implication @avikivity. This PR creates an impression that it just removes features that has been always on for many releases but this is not true for MC_SSTABLE_FORMAT. It is entierly possible that some users have decided to stay on LA sstables. Removing the MC_SSTABLE_FORMAT feature removes this ability. It is at least something that should be mentioned in the release notes (@tzach ). The question also is whether this is a conscious decision or an accidental result.
> > 
> > 
> > Why is removing this feature removing this ability? The feature merely means that the cluster is _capable_ of using MC and says nothing about whether it is actually using it.
> 
> Currently users may decide to write LA sstables in their clusters. I don't know why they would do that but it's certainly an option. After this PR all sstables will be written as MC. Ability of reading LA sstables is not affected.

The point is that since we only support upgrades from a previous release, if users uses LA, then they already have explicitly disabled MC in config, right? So nothing will change for them during the upgrade to this state of the code.

haaawk commented:
> > > > > Two of those features (MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES) are controlled by a flag. Have we removed the flag to make it impossible to turn off those features?
> > > > 
> > > > 
> > > > The flag exists, but is no longer consulted.
> > > 
> > > 
> > > There's a deeper implication. This PR creates an impression that it just removes features that has been always on for many releases but this is not true for MC_SSTABLE_FORMAT. It is entierly possible that some users have decided to stay on LA sstables. Removing the MC_SSTABLE_FORMAT feature removes this ability. It is at least something that should be mentioned in the release notes (@tzach ). The question also is whether this is a conscious decision or an accidental result.
> > 
> > 
> > Actually, removing those MC_SSTABLE_FORMAT is just unsafe. In the rolling upgrade when cluster has MC_SSTABLE_FORMAT disabled, we may end up in a state where some nodes use LA and others MC. When we add UNBOUNDED_RANGE_TOMBSTONES that can't be handled by LA we will end up with troubles. Maybe the unbounded tombstone will be lost or maybe it will be wrongly persisted. I don't know.
> > The correct way to remove MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES is to remove the flag that controls them in one release and only in the next remove those features. Or am I missing something @avikivity ?
> 
> We should never be able to upgrade from a cluster that doesn't have MC_SSTABLE_FORMAT enabled **and** this series is applied, because we only support direct upgrades from the previous release, not really old ones. That's why we only deprecate features that are over 2 years old.

User can have MC_SSTABLE_FORMAT disabled by a flag for all this time. When you upgrade there's no MC_SSTABLE_FORMAT feature any more so we won't even check that during rolling upgrade, will we?

haaawk commented:
> > > > > > Two of those features (MC_SSTABLE_FORMAT and UNBOUNDED_RANGE_TOMBSTONES) are controlled by a flag. Have we removed the flag to make it impossible to turn off those features?
> > > > > 
> > > > > 
> > > > > The flag exists, but is no longer consulted.
> > > > 
> > > > 
> > > > There's a deeper implication @avikivity. This PR creates an impression that it just removes features that has been always on for many releases but this is not true for MC_SSTABLE_FORMAT. It is entierly possible that some users have decided to stay on LA sstables. Removing the MC_SSTABLE_FORMAT feature removes this ability. It is at least something that should be mentioned in the release notes (@tzach ). The question also is whether this is a conscious decision or an accidental result.
> > > 
> > > 
> > > Why is removing this feature removing this ability? The feature merely means that the cluster is _capable_ of using MC and says nothing about whether it is actually using it.
> > 
> > 
> > Currently users may decide to write LA sstables in their clusters. I don't know why they would do that but it's certainly an option. After this PR all sstables will be written as MC. Ability of reading LA sstables is not affected.
> 
> The point is that since we only support upgrades from a previous release, if users use LA, then they already have explicitly disabled MC in config, right? So nothing will change for them during the upgrade to this state of the code.

If user uses LA then MC_SSTABLE_FORMAT is disabled. After this PR MC_SSTABLE_FORMAT is alway enabled.

psarna commented:
@avikivity Ok, so after further discussion with @haaawk it looks like the problem is as follows:
1. Users are able to effectively block the MC_SSTABLE_FORMAT feature by disabling it in config
2. The UNBOUNDED_RANGE_TOMBSTONES feature depends on MC
3. Unbounded range tombstones are not supported for LA sstables and should never be sent to nodes which use LA

Now, if we have a cluster with mc format disabled in config and we upgrade to a new release with this series applied, we can approach a situation where some nodes usee LA sstables, but at the same time the new nodes **assume** that it's ok to write unbounded range tombstones, because the feature is assumed to be true. And here's our problem, since LA sstables can't digest these tombstones.

psarna commented:
> @avikivity Ok, so after further discussion with @haaawk it looks like the problem is as follows:
> 
>     1. Users are able to effectively block the MC_SSTABLE_FORMAT feature by disabling it in config
> 
>     2. The UNBOUNDED_RANGE_TOMBSTONES feature depends on MC
> 
>     3. Unbounded range tombstones are not supported for LA sstables and should never be sent to nodes which use LA
> 
> 
> Now, if we have a cluster with mc format disabled in config and we upgrade to a new release with this series applied, we can approach a situation where some nodes usee LA sstables, but at the same time the new nodes **assume** that it's ok to write unbounded range tombstones, because the feature is assumed to be true. And here's our problem, since LA sstables can't digest these tombstones.

Aaand that's how I understand the problem, but @haaawk is checking one more thing and will either support this thesis or debunk it. I'm waiting for input.

haaawk commented:
> > @avikivity Ok, so after further discussion with @haaawk it looks like the problem is as follows:
> > ```
> > 1. Users are able to effectively block the MC_SSTABLE_FORMAT feature by disabling it in config
> > 
> > 2. The UNBOUNDED_RANGE_TOMBSTONES feature depends on MC
> > 
> > 3. Unbounded range tombstones are not supported for LA sstables and should never be sent to nodes which use LA
> > ```
> > 
> > 
> > Now, if we have a cluster with mc format disabled in config and we upgrade to a new release with this series applied, we can approach a situation where some nodes usee LA sstables, but at the same time the new nodes **assume** that it's ok to write unbounded range tombstones, because the feature is assumed to be true. And here's our problem, since LA sstables can't digest these tombstones.
> 
> Aaand that's how I understand the problem, but @haaawk is checking one more thing and will either support this thesis or debunk it. I'm waiting for input.

It seems that it was possible to dissable MC before this PR:
https://github.com/scylladb/scylla/blob/10364fca6ee41aa1697edfc80e67d08ce6c26137/gms/feature_service.cc#L104

haaawk commented:
> > > @avikivity Ok, so after further discussion with @haaawk it looks like the problem is as follows:
> > > ```
> > > 1. Users are able to effectively block the MC_SSTABLE_FORMAT feature by disabling it in config
> > > 
> > > 2. The UNBOUNDED_RANGE_TOMBSTONES feature depends on MC
> > > 
> > > 3. Unbounded range tombstones are not supported for LA sstables and should never be sent to nodes which use LA
> > > ```
> > > 
> > > 
> > > Now, if we have a cluster with mc format disabled in config and we upgrade to a new release with this series applied, we can approach a situation where some nodes usee LA sstables, but at the same time the new nodes **assume** that it's ok to write unbounded range tombstones, because the feature is assumed to be true. And here's our problem, since LA sstables can't digest these tombstones.
> > 
> > 
> > Aaand that's how I understand the problem, but @haaawk is checking one more thing and will either support this thesis or debunk it. I'm waiting for input.
> 
> It seems that it was possible to disable MC before this PR:
> https://github.com/scylladb/scylla/blob/10364fca6ee41aa1697edfc80e67d08ce6c26137/gms/feature_service.cc#L104

BTW, it is safe to remove MC_SSTABLE_FORMAT but UNBOUNDED_RANGE_TOMBSTONES probably needs to wait at least one more release.

psarna commented:
I self-dequeued the series due to the risk associated with removing `UNBOUNDED_RANGE_TOMBSTONES`.
v2:
 - removed removing `UNBOUNDED_RANGE_TOMBSTONES`, it will wait for at least one more release cycle until it's purged

psarna commented:
* updated the description to no longer mention unbounded range tombstones

haaawk commented on db/sstables-format-selector.cc:
```diff
@@ -75,7 +74,6 @@ future<> sstables_format_selector::do_maybe_select_format(sstables::sstable_vers
 future<> sstables_format_selector::start() {
     assert(this_shard_id() == 0);
     return read_sstables_format().then([this] {
-        _features.local().cluster_supports_mc_sstable().when_enabled(_mc_feature_listener);
```
I think this might be unsafe. UNBOUNDED_RANGE_TOMBSTONE is being enabled by this listener. I believe that you you need to enable UNBOUNDED_RANGE_TOMBSTONE by default if you're removing this listener.

haaawk replied:
Probably just removing this line will do the job:
https://github.com/scylladb/scylla/blob/10364fca6ee41aa1697edfc80e67d08ce6c26137/gms/feature_service.cc#L100

haaawk replied:
Another question is whether we can remove _mc_feature_listener althogether?

psarna replied:
It doesn't really get enabled by this listener - in this series, `_selected_format` defaults to `mc`, so this feature is masked and then unmasked unconditionally. This masking-and-then-unmasking makes little sense and it should be purged, but it should nonetheless work fine.

And I did remove _mc_feature_listener altogether in this series, because the `_selected_format` now defaults to mc. Whether it makes sense or not is a separate issue and I should at least mention it in the commit message though.

psarna replied:
My main source of confusion was that I assumed that `MC_SSTABLE_FORMAT` feature was only used to announce that a node **can** support mc sstables, while in reality it's also used to actually decide if mc format is used or not. That complicates the whole business with deprecating features.

psarna commented:
I think I'll actually go one step further and drop removing `MC_SSTABLE_FORMAT` feature from this series as well. It sounds risky to just remove it, since this feature's presence is controlled by a flag controlled by users via config, which makes the feature non-standard and the assumption that upgrades are safe because the 2 year window is large does not apply here.

psarna commented:
v3:
 - dropped dropping MC format feature as well due to too many uncertainties - it can be reconsidered in a future follow-up

haaawk reviewed (approved):
LGTM

psarna commented:
un ping

psarna commented:
Deprecated because some dtests depend on detecting row level repair by grepping logs, and once the feature got deprecated the log message is no longer there. I'll update dtests first and ping after I'm done.

psarna commented:
Once https://github.com/scylladb/scylla-dtest/pull/2033 is merged, this series will be ready for readmission.

psarna commented:
Rebased; scylladb/scylla-dtest#2033 is merged to dtest. Ready for requeuing.

avikivity commented on gms/feature_service.cc:
```diff
@@ -43,12 +43,12 @@ constexpr std::string_view features::XXHASH = "XXHASH";
 constexpr std::string_view features::ROLES = "ROLES";
 constexpr std::string_view features::LA_SSTABLE = "LA_SSTABLE_FORMAT";
 constexpr std::string_view features::STREAM_WITH_RPC_STREAM = "STREAM_WITH_RPC_STREAM";
```
I wouldn't say the feature is deprecated, in fact it's being made mandatory.

I understand what is meant, but it's better to be more precise.

psarna commented:
v3:
 - removed helper variables from sstable writer configs, because they are now always true
 - as for the wording of "deprecate" - I'll change the title of this PR to "Clean up (...)", since that's how the previous similar series was named

avikivity commented:
Individual patches still talk about deprecating features.

psarna commented:
v4:
 - reworded all commit messages to emphasize that the features are not themselves going away, but they are just assumed to be always available; grep shows 0 occurrences of the words "deprecate" and "drop" now

psarna commented:
(the last occurrence of "deprecate" can be spotted in my local branch name, but I'd rather not change that one due to how pull requests work)
